### PR TITLE
Verbose display of exceptions [ECR-1035].

### PR DESCRIPTION
**Pull request by alexander-irbis**
## Overview

Fixed display of exception messages and stack traces.
These was broken with earlier versions of jni-rs.

---
See: https://jira.bf.local/browse/ECR-1035

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated tests
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] The [continuous integration build](https://www.travis-ci.com/exonum/exonum-java-binding) passes

### CODE REVIEW

File: exonum-java-binding-core/rust/src/utils/errors.rs [hunk 1]
```diff
@@ -97,11 +97,14 @@ pub fn get_and_clear_java_exception<'e>(env: &'e JNIEnv) -> JObject<'e> {
...
     assert!(!exception.is_null(), "No exception thrown.");
-    format!(
-        "Java exception: {}; {:?}\n{}{:#?}",
-        unwrap_jni_verbose(env, get_class_name(env, exception)),
-        unwrap_jni_verbose(env, get_exception_message(env, exception)),
-        unwrap_jni_verbose(env, get_exception_stack_trace(env, exception)),
-        jni_error.backtrace(),
-    )
+    let format = || {
+        Ok(format!(
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/rust/src/utils/errors.rs]: I'm against using such implementation on 'happy' paths, where an exception is an expected behaviour (e.g., if a transaction message is malformed, the service *must* throw an exception). I'd restrict usage of this method to actual errors.

On top of that, the exceptions, including errors, are logged on Java side, with proper handling of various stack traces (causes & suppressions).

File: exonum-java-binding-core/rust/src/utils/jni.rs [hunk 2]
```diff
@@ -13,50 +13,42 @@ pub fn get_class_name(env: &JNIEnv, object: JObject) -> JniResult<String> {
     convert_to_string(env, class_name)
 }
 
-pub fn get_exception_message(_env: &JNIEnv, exception: JObject) -> JniResult<String> {
+/// Returns the message from the exception if it is not null.
+///
+/// `exception` should extend `java.lang.Throwable` and be not null
+pub fn get_exception_message(env: &JNIEnv, exception: JObject) -> JniResult<Option<String>> {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/rust/src/utils/jni.rs]: I'd add a couple of tests for these methods.

File: exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs [hunk 3]
```diff
@@ -146,6 +148,42 @@ fn get_and_clear_java_exception_if_exception_occurred() {
         .unwrap();
 }
 
+// FIXME blocked by https://jira.bf.local/browse/ECR-1998
+// Currently (jni 0.11) there is no way to throw exception from native code
+// without message argument.
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs]: Isn't it possible to 
1. Create an exception using no-arg constructor using env.new_object?
2. Throw it with that method?

**Comment by dmitry-timofeev** [comment]
> I'm against using such implementation on 'happy' paths, where an exception is an expected behaviour (e.g., if a transaction message is malformed, the service must throw an exception). I'd restrict usage of this method to actual errors.

> On top of that, the exceptions, including errors, are logged on Java side, with proper handling of various stack traces (causes & suppressions).

That still applies — I don't think the current implementation shall be used in either build in `check_error_on_exception`.

File: exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs [hunk 4]
```diff
@@ -157,8 +186,14 @@ fn get_and_clear_java_exception_if_no_exception_occurred() {
         .unwrap();
 }
 
-fn throw(env: &JNIEnv, e: &str) -> JniResult<()> {
-    env.throw((e, ""))?;
+fn throw(env: &JNIEnv, exception_class: &str) -> JniResult<()> {
+    // FIXME throw an exception without a stub message https://jira.bf.local/browse/ECR-1998
+    env.throw((exception_class, ""))?;
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs]: Would it be possible to

1 Create an exception using no-arg constructor using env.new_object?
2 Throw it with env.throw?

Comment (alexander-irbis) [reply exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs]: 2. Currently it is impossible with jni-rs 😞

Comment (dmitry-timofeev) [reply exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs]: That's weird, I thought a reference to a throwable object works with signature `Desc<Throwable>`: https://docs.rs/jni/0.10.1/jni/struct.JNIEnv.html#method.throw

Comment (alexander-irbis) [reply exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs]: I found implementations of Desc<JThrowable> only here: https://github.com/prevoty/jni-rs/blob/master/src/wrapper/descriptors/exception_desc.rs

All of them imply a pair of:
- an object with a class name or `java/lang/Exception` as a default, and
- an argument, castable to a string.

Comment (dmitry-timofeev) [reply exonum-java-binding-core/rust/integration_tests/tests/utils_errors.rs]: I see, thanks!

**Pull request merged**